### PR TITLE
performance: replace a critical usage of Array.concat with for loop/push

### PR DESCRIPTION
**Pull request by natesilva**
replace a critical usage of `Array.concat` with a `for` loop and `Array.prototype.push` for a performance improvement on large queries

### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

_NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._

### Description of change

This change provides a performance boost to large queries. I found this when profiling an ugly query that returns tens of thousands of rows, with `include`s 4 levels deep. The line with `Array.concat()` is called hundreds of thousands of times and is a bottleneck.

For loops are [much faster](https://jsperf.com/array-concat-vs-push-sequelize) and in my real-world testing with the ugly query, was about 3 times faster, going from 60-90s down to about 21s.

Attached: part of the `node --profile` output that pointed me to this section of code. Profiling was based on Sequelize 4.0.0-2. [profiling-partial.txt](https://github.com/sequelize/sequelize/files/743526/profiling-partial.txt)

\* `npm run test` passes except for two tests that I don’t think are related (login information and incorrect credentials)

### CODE REVIEW

**Comment by sushantdhiman** [comment]
@natesilva Could you please compare results using https://github.com/sushantdhiman/sequelize-benchmark , It will help us decide if there is any significant performance improvement and how much.

**Comment by natesilva** [comment]
I haven’t been able to run the benchmark. With MySQL it reports a `Got a packet bigger than 'max_allowed_packet' bytes`; with PostgreSQL it never seems to finish. (This isn’t a problem in my PR, it does this even against the Sequelize master branch.) So my next step is to debug the benchmark.

**Comment by sushantdhiman** [comment]
As tested with [sequelize-benchmark@1.1.0](https://github.com/sushantdhiman/sequelize-benchmark)
**PR**
```
▶ DIALECT=postgres SEQ_PORT=8998 SEQ_USER=sequelize_test SEQ_PW=sequelize_test sequelize-benchmark
  sequelize:perf Using postgres DIALECT. +0ms
  sequelize:perf Env ready +285ms
  sequelize:perf Starting findAll +15s
  sequelize:perf finding 1500 rows x 2.19 ops/sec ±1.04% (15 runs sampled) (mean 0.457374s) +7s
  sequelize:perf finding(raw) 1500 rows x 2.27 ops/sec ±0.82% (16 runs sampled) (mean 0.440951s) +7s
  sequelize:perf finding(plain) 1500 rows x 2.17 ops/sec ±0.44% (15 runs sampled) (mean 0.460980s) +7s
  sequelize:perf finding 1500 rows with join x 1.98 ops/sec ±1.28% (14 runs sampled) (mean 0.503912s) +7s
  sequelize:perf finding 1500 rows with join (heavy includes) x 1.74 ops/sec ±1.63% (13 runs sampled) (mean 0.573157s) +8s
  sequelize:perf finding 10 rows x 131 ops/sec ±0.82% (77 runs sampled) (mean 0.007630s) +6s
  sequelize:perf finding(raw) 10 rows x 140 ops/sec ±0.40% (80 runs sampled) (mean 0.007146s) +6s
  sequelize:perf finding(plain) 10 rows x 132 ops/sec ±0.49% (84 runs sampled) (mean 0.007568s) +6s
  sequelize:perf finding 10 rows with join x 120 ops/sec ±0.60% (79 runs sampled) (mean 0.008320s) +6s
  sequelize:perf Starting bulkCreate +159ms
  sequelize:perf inserting 10000 rows x 0.53 ops/sec ±0.51% (7 runs sampled) (mean 1.873272s) +13s
  sequelize:perf inserting 1000 rows x 8.63 ops/sec ±2.21% (45 runs sampled) (mean 0.115810s) +6s
  sequelize:perf inserting 10 rows x 153 ops/sec ±3.15% (72 runs sampled) (mean 0.006547s) +6s
  sequelize:perf Benchmark finished +0ms

▶ DIALECT=sqlite SEQ_PORT=8998 SEQ_USER=sequelize_test SEQ_PW=sequelize_test sequelize-benchmark  
  sequelize:perf Using sqlite DIALECT. +0ms
  sequelize:perf Env ready +142ms
  sequelize:perf Starting findAll +6s
  sequelize:perf finding 1500 rows x 20.45 ops/sec ±2.63% (52 runs sampled) (mean 0.048907s) +6s
  sequelize:perf finding(raw) 1500 rows x 46.46 ops/sec ±1.16% (73 runs sampled) (mean 0.021523s) +6s
  sequelize:perf finding(plain) 1500 rows x 20.38 ops/sec ±2.63% (51 runs sampled) (mean 0.049071s) +6s
  sequelize:perf finding 1500 rows with join x 9.18 ops/sec ±2.32% (47 runs sampled) (mean 0.108895s) +6s
  sequelize:perf finding 1500 rows with join (heavy includes) x 6.48 ops/sec ±2.16% (35 runs sampled) (mean 0.154396s) +6s
  sequelize:perf finding 10 rows x 216 ops/sec ±1.48% (79 runs sampled) (mean 0.004632s) +6s
  sequelize:perf finding(raw) 10 rows x 232 ops/sec ±0.71% (82 runs sampled) (mean 0.004314s) +6s
  sequelize:perf finding(plain) 10 rows x 219 ops/sec ±1.07% (80 runs sampled) (mean 0.004561s) +6s
  sequelize:perf finding 10 rows with join x 170 ops/sec ±1.77% (80 runs sampled) (mean 0.005880s) +6s
  sequelize:perf Starting bulkCreate +48ms
  sequelize:perf inserting 10000 rows x 1.28 ops/sec ±2.49% (11 runs sampled) (mean 0.780092s) +9s
  sequelize:perf inserting 1000 rows x 12.03 ops/sec ±2.01% (59 runs sampled) (mean 0.083145s) +6s
  sequelize:perf inserting 10 rows x 384 ops/sec ±1.98% (77 runs sampled) (mean 0.002605s) +6s
  sequelize:perf Benchmark finished +0ms
```

**Master**

``` 
▶ DIALECT=sqlite SEQ_PORT=8998 SEQ_USER=sequelize_test SEQ_PW=sequelize_test sequelize-benchmark  
  sequelize:perf Using sqlite DIALECT. +0ms
  sequelize:perf Env ready +136ms
  sequelize:perf Starting findAll +6s
  sequelize:perf finding 1500 rows x 20.64 ops/sec ±4.37% (53 runs sampled) (mean 0.048439s) +6s
  sequelize:perf finding(raw) 1500 rows x 46.48 ops/sec ±0.78% (73 runs sampled) (mean 0.021515s) +6s
  sequelize:perf finding(plain) 1500 rows x 21.31 ops/sec ±1.52% (53 runs sampled) (mean 0.046932s) +6s
  sequelize:perf finding 1500 rows with join x 8.54 ops/sec ±3.26% (45 runs sampled) (mean 0.117158s) +6s
  sequelize:perf finding 1500 rows with join (heavy includes) x 5.98 ops/sec ±2.32% (33 runs sampled) (mean 0.167194s) +6s
  sequelize:perf finding 10 rows x 226 ops/sec ±1.13% (80 runs sampled) (mean 0.004419s) +6s
  sequelize:perf finding(raw) 10 rows x 246 ops/sec ±1.05% (81 runs sampled) (mean 0.004062s) +6s
  sequelize:perf finding(plain) 10 rows x 222 ops/sec ±1.44% (81 runs sampled) (mean 0.004498s) +6s
  sequelize:perf finding 10 rows with join x 180 ops/sec ±1.35% (82 runs sampled) (mean 0.005549s) +6s
  sequelize:perf Starting bulkCreate +43ms
  sequelize:perf inserting 10000 rows x 1.28 ops/sec ±2.44% (11 runs sampled) (mean 0.782450s) +9s
  sequelize:perf inserting 1000 rows x 12.03 ops/sec ±2.26% (59 runs sampled) (mean 0.083100s) +6s
  sequelize:perf inserting 10 rows x 386 ops/sec ±2.15% (78 runs sampled) (mean 0.002594s) +6s
  sequelize:perf Benchmark finished +1ms
 
▶ DIALECT=postgres SEQ_PORT=8998 SEQ_USER=sequelize_test SEQ_PW=sequelize_test sequelize-benchmark
  sequelize:perf Using postgres DIALECT. +0ms
  sequelize:perf Env ready +297ms
  sequelize:perf Starting findAll +15s
  sequelize:perf finding 1500 rows x 2.13 ops/sec ±1.51% (15 runs sampled) (mean 0.468794s) +7s
  sequelize:perf finding(raw) 1500 rows x 2.21 ops/sec ±3.97% (15 runs sampled) (mean 0.452262s) +7s
  sequelize:perf finding(plain) 1500 rows x 2.16 ops/sec ±0.38% (15 runs sampled) (mean 0.462087s) +7s
  sequelize:perf finding 1500 rows with join x 1.95 ops/sec ±1.86% (14 runs sampled) (mean 0.513363s) +7s
  sequelize:perf finding 1500 rows with join (heavy includes) x 1.77 ops/sec ±1.51% (13 runs sampled) (mean 0.565629s) +8s
  sequelize:perf finding 10 rows x 130 ops/sec ±0.91% (77 runs sampled) (mean 0.007711s) +6s
  sequelize:perf finding(raw) 10 rows x 138 ops/sec ±0.67% (80 runs sampled) (mean 0.007241s) +6s
  sequelize:perf finding(plain) 10 rows x 130 ops/sec ±0.84% (81 runs sampled) (mean 0.007697s) +6s
  sequelize:perf finding 10 rows with join x 119 ops/sec ±0.66% (79 runs sampled) (mean 0.008395s) +6s
  sequelize:perf Starting bulkCreate +131ms
  sequelize:perf inserting 10000 rows x 0.53 ops/sec ±0.64% (7 runs sampled) (mean 1.897704s) +13s
  sequelize:perf inserting 1000 rows x 8.71 ops/sec ±2.12% (45 runs sampled) (mean 0.114747s) +6s
  sequelize:perf inserting 10 rows x 153 ops/sec ±3.09% (72 runs sampled) (mean 0.006545s) +6s
  sequelize:perf Benchmark finished +1ms
```

There are slight improvements, benchmark currently doesn't support nested includes where this performance improvement will be mostly applicable.

Overall I think we should go ahead and merge it

File: changelog.md [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # Future
+- [PERFORMANCE] more efficient array handing for certain large queries [#7175](https://github.com/sequelize/sequelize/pull/7175)
```
Comment (sushantdhiman) [inline changelog.md]: Efficient handling of nested associations arrays

File: lib/model.js [hunk 2]
```diff
@@ -1578,7 +1578,10 @@ class Model {
             // Force array so we can concat no matter if it's 1:1 or :M
             if (!Array.isArray(associations)) associations = [associations];
 
-            return memo.concat(associations);
+            for (let i = 0, len = associations.length; i !== len; ++i) {
+              memo.push(associations[i]);
+            }
+            return memo;
```
Comment (sushantdhiman) [inline lib/model.js]: New line before `return`

**Comment by janmeier** [comment]
Thanks for the thorough review work @sushantdhiman , and thanks for taking the time to investigate this @natesilva ! :+1:

**Pull request merged**